ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -470,9 +470,30 @@ private void releaseResource() {
     }
   }
 
+  /**
+   * Release the blob content of chunks, including building chunk.
+   */
+  private synchronized void releaseDataForAllChunks() {
+    for (PutChunk chunk : putChunks) {
+      if (chunk.isBuilding() || chunk.isReady() || chunk.isComplete()) {
```

Comment:
Do we need to handle the encrypting state as well? Is there harm to just releasing all chunks instead of having this if condition.

Reply:
encrypted chunk has its owner thread to release the data. There might be some other race condition there. Let me take another look on the encrypted chunk and add a test case for this code.

Reply:
It does look like the EncryptJob takes in `buf.retainedDuplicate()` though and handles releasing that. Do we also have to release the original `buf`?

Reply:
we release the original `buf` in the encryptionCallback method. The first if statement checks if the chunk is data chunk, if it's, we release the `buf` no matter if there is an exception.